benchmarks/src/main/java/org/elasticsearch/benchmark/indices/resolution/IndexNameExpressionResolverBenchmark.java

```diff
@@ -0,0 +1,137 @@
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+@State(Scope.Benchmark)
+@Fork(3)
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+@SuppressWarnings("unused") // invoked by benchmarking framework
+public class IndexNameExpressionResolverBenchmark {
+
+    private static final String DATA_STREAM_PREFIX = "my-ds-";
+    private static final String INDEX_PREFIX = "my-index-";
+
+    @Param(
+        {
+            // # data streams | # indices
+            "              1000|        100",
+            "              5000|        500",
+            "             10000|       1000" }
+    )
+    public String resourceMix = "100|10";
+
+    @Setup
+    public void setUp() {
+        final String[] params = resourceMix.split("\\|");
+
+        int numDataStreams = toInt(params[0]);
+        int numIndices = toInt(params[1]);
+
+        Metadata.Builder mb = Metadata.builder();
+        indices = new String[numIndices + numDataStreams * (numIndices + 1)];
+        int position = 0;
+        for (int i = 1; i <= numIndices; i++) {
+            String indexName = INDEX_PREFIX + i;
+            createIndexMetadata(indexName, mb);
+            indices[position++] = indexName;
+        }
+
+        for (int i = 1; i <= numDataStreams; i++) {
+            String dataStreamName = DATA_STREAM_PREFIX + i;
+            List<Index> backingIndices = new ArrayList<>();
+            for (int j = 1; j <= numIndices; j++) {
+                String backingIndexName = DataStream.getDefaultBackingIndexName(dataStreamName, j);
+                backingIndices.add(createIndexMetadata(backingIndexName, mb).getIndex());
+                indices[position++] = backingIndexName;
+            }
+            indices[position++] = dataStreamName;
+            mb.put(DataStream.builder(dataStreamName, backingIndices).build());
+        }
+
+        clusterState = ClusterState.builder(ClusterName.DEFAULT).metadata(mb).build();
+        resolver = new IndexNameExpressionResolver(new ThreadContext(Settings.EMPTY), new SystemIndices(List.of()));
+    }
+
+    private IndexMetadata createIndexMetadata(String indexName, Metadata.Builder mb) {
+        IndexMetadata indexMetadata = IndexMetadata.builder(indexName)
+            .settings(Settings.builder().put(IndexMetadata.SETTING_VERSION_CREATED, IndexVersion.current()))
+            .numberOfShards(1)
+            .numberOfReplicas(0)
+            .build();
+        mb.put(indexMetadata, false);
+        return indexMetadata;
+    }
+
+    private IndexNameExpressionResolver resolver;
+    private ClusterState clusterState;
+    private String[] indices;
+
+    @Benchmark
+    public String[] resolveResourcesListToConcreteIndices() {
+        return resolver.concreteIndexNames(clusterState, new IndicesRequest() {
```

Comment:
This is a bit of a technicality but lets make this more realistic and make the `IndicesRequest` implementation here a `static` named class and have it copy the indices list in it in the constructor? :)